File: 0643. Maximum Average Subarray I/0643.java
// time complexity: O(n*k)
// space complexity: O(1)
class Solution {
    public double findMaxAverage(int[] nums, int k) {
        int i = 0;
        int total = Integer.MIN_VALUE;

        // loop until the window reaches the end of the array
        while (i + k <= nums.length) {  // runs approximately O(n) times
            int current_total = 0;
            // sum up the values within the current window of size k
            for (int m = i; m < i + k; m++) {  // each iteration takes O(k) time
                current_total += nums[m];
            }
            // update total if the current window sum is larger
            total = Math.max(total, current_total);
            i++;
        }
        // return the maximum average by dividing the maximum total sum by k
        return (double) total / k;
    }
}

File: 0643. Maximum Average Subarray I/0643.py
# time complexity: O(n)
# space complexity: O(1)
class Solution(object):
    def findMaxAverage(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: float
        """
        # calculate the sum of the first window of size k
        window_sum = sum(nums[:k])
        max_sum = window_sum
        
        # slide the window through the array, updating the sum in O(1) time per step
        for i in range(k, len(nums)):
            # remove the element leaving the window and add the new element
            window_sum += nums[i] - nums[i - k]
            # update max_sum if the current window sum is greater
            if window_sum > max_sum:
                max_sum = window_sum
        
        # return the maximum average by dividing the maximum sum by k
        return float(max_sum) / k